docarray/base_document/document.py

```diff
@@ -79,6 +84,128 @@ class MyDocument(BaseDocument):
 
         :param other: The Document with which to update the contents of this
         """
-        from docarray.utils.reduce import reduce_docs
-
-        reduce_docs(self, other)
+        if type(self) != type(other):
+            raise Exception(
+                f'Update operation can only be applied to '
+                f'Documents of the same type. '
+                f'Trying to update Document of type '
+                f'{type(self)} with Document of type '
+                f'{type(other)}'
+            )
+        from docarray.utils.reduce import reduce
+        from docarray import DocumentArray
+
+        from collections import namedtuple
+
+        # Declaring namedtuple()
+        _FieldGroups = namedtuple(
+            '_FieldGroups',
+            [
+                'simple_non_empty_fields',
+                'list_fields',
+                'set_fields',
+                'dict_fields',
+                'nested_docarray_fields',
+                'nested_docs_fields',
+            ],
+        )
+
+        FORBIDDEN_FIELDS_TO_UPDATE = ['ID']
+
+        def _group_fields(doc: 'BaseDocument') -> _FieldGroups:
+            simple_non_empty_fields: List[str] = []
+            list_fields: List[str] = []
+            set_fields: List[str] = []
+            dict_fields: List[str] = []
+            nested_docs_fields: List[str] = []
+            nested_docarray_fields: List[str] = []
+
+            for field_name, field in doc.__fields__.items():
+                if field_name not in FORBIDDEN_FIELDS_TO_UPDATE:
+                    field_type = doc._get_field_type(field_name)
+                    if not isinstance(field_type, _GenericAlias) and issubclass(
+                        field_type, DocumentArray
+                    ):
+                        nested_docarray_fields.append(field_name)
+                    elif (
+                        isinstance(field_type, _GenericAlias)
+                        and field_type.__origin__ is list
+                    ):
```

Comment:
what Python version are u referring to?

Reply:
in python 3.9 you can do

```python
class A:
   b: list[inŧ]
```

instead of 

```python
class A:
   b: List[int]
```

I am almost sure they have maintain compatibility with older version but the fact that we are using a private object from the typing library tell me that we are not 100% sure we will support this new way.

Maybe there is a way not to use the `_GenericAlias` since it is private ?

Reply:
Oh actually we can use 

```python
from typing_inspect import is_generic_type
```

 we already rely on this library for some type hint things and they make sure that their function are compatible with every python version


https://github.com/ilevkivskyi/typing_inspect

Reply:
can u propose a PR on top of mine to do this? I am not sure what u are intending to tell. I did trial and error until I found the working formula

Reply:
I pushed a commit